lib/api/decorators/schema.rb

```diff
@@ -0,0 +1,128 @@
+                                                                    name: title,
+                                                                    required: required,
+                                                                    writable: writable)
+          schema.min_length = min_length if min_length
+          schema.max_length = max_length if max_length
+
+          property property,
+                   getter: -> (*) { schema },
+                   writeable: false
+        end
+
+        def schema_with_allowed_link(property,
+                                          type: make_type(property),
+                                          title: make_title(property),
+                                          href_callback:,
+                                          required: true,
+                                          writable: true)
+          raise ArgumentError if property.nil?
+
+          property property,
+                   exec_context: :decorator,
+                   getter: -> (*) {
+                     representer = ::API::Decorators::AllowedValuesByLinkRepresenter.new(
+                       type: type,
+                       name: title,
+                       required: required,
+                       writable: writable)
+
+                     if represented.defines_assignable_values?
+                       representer.allowed_values_href = instance_eval(&href_callback)
+                     end
+
+                     representer
+                   }
+        end
+
+        def schema_with_allowed_collection(property,
+                                                type: make_type(property),
+                                                title: make_title(property),
+                                                values_callback:,
+                                                value_representer:,
+                                                link_factory:,
+                                                required: true,
+                                                writable: true)
+          raise ArgumentError unless property
+
+          property property,
+                   exec_context: :decorator,
+                   getter: -> (*) {
+                     representer = ::API::Decorators::AllowedValuesByCollectionRepresenter.new(
+                       type: type,
+                       name: title,
+                       current_user: current_user,
+                       value_representer: value_representer,
+                       link_factory: -> (value) { instance_exec(value, &link_factory) },
+                       required: required,
+                       writable: writable)
+
+                     if represented.defines_assignable_values?
+                       representer.allowed_values = instance_exec(&values_callback)
+                     end
+
+                     representer
+                   }
+        end
```

Comment:
I'll have another look at it, but I still doubt it is as easy...

## Problems

**Assignment of allowed values different for both**

You say the shared part contains the assignment of the allowed values:
````ruby
if represented.defines_assignable_values?
  representer.allowed_values = instance_exec(&values_callback)
end
````

However, strictly speaking the code is not identical. On one case a method `allowed_values` is called, in the other case the method is called `allowed_values_href`. That is because the datatype of both calls is significantly different (list of things vs. string).

*Possible workaround:* Calling both methods the same, and expecting the caller to provide the correct result in the callback.

**Need for instance evaluation in representer**

The representers of `schema_with_allowed_collection` are created using `instance_exec` for their `link_factory`. I think the only reason for that is that we want to use `api_v3_paths`, which is only available in the context of the representer, but not in the context of its class.

*Possible workaround:* Deviate from the usual method of including the `PathHelper`.
Option 1: I could reference `ApiV3Path` directly instead of `api_v3_paths`, probably inside a class method of my own.
Option 2: It might be possible to include the `PathHelper` module within my `class << self`. My Ruby foo is not good enough to know if this will work, but I would try it. However I still don't know if this would be a good or a bad idea.

**More need for representer context**

After another look at the representer created in `schema_with_allowed_collection`, I noticed that there is another need for code evaluation inside the representer context (as opposed to execution on the class level): We access the instance method called `current_user`, which is important to correctly create the embedded instances of the allowed values collection.

I don't know if it is strictly neccessary to have the `current_user` as instance variable of the representers. However, this si a design decision that was made before I arrived here and I think it would at least take some effort to remove the `current_user` instance method from all representers (which is what we needed to do consequently).

*Possible workaround:* See above, I don't know if this was feasible.

*Generic workaround for both problems above:* Add another callback that creates the representer for allowed values and also evaluate that one in the right context. But I doubt, that readability is something that can be improved by adding more layers of indirection.

**tl;dr:**

I think our main problem is that we need the context of the schema representer to create the allowed value representers.